### PR TITLE
LibWeb: Implement document.execCommand("insertParagraph")

### PR DESCRIPTION
**Pull request by gmta**
At least 1045 additional WPT passes in `editing`!

### CODE REVIEW

File: Libraries/LibWeb/Editing/Commands.cpp [hunk 1]
```diff
@@ -384,9 +368,347 @@ bool command_delete_action(DOM::Document& document, String const&)
...
+            descendant->remove_attribute(HTML::AttributeNames::id);
+        return TraversalDecision::Continue;
+    });
+
+    // 29. Call appendChild(frag) on new container.
+    MUST(new_container->append_child(frag));
+
+    // 30. While container's lastChild is a prohibited paragraph child, set container to its lastChild.
+    while (container->last_child() && is_prohibited_paragraph_child(*container->last_child()))
+        container = container->last_child();
```
Comment (tcl3) [inline Libraries/LibWeb/Editing/Commands.cpp:649]: Could `last_child()` return null here?

File: Libraries/LibWeb/Editing/Commands.cpp [hunk 2]
```diff
@@ -384,9 +368,347 @@ bool command_delete_action(DOM::Document& document, String const&)
     return true;
 }
 
+bool command_insert_paragraph_action(DOM::Document& document, String const&)
```
Comment (tcl3) [inline Libraries/LibWeb/Editing/Commands.cpp:371]: Can we add a spec link here, please?

File: Libraries/LibWeb/Editing/Internal/Algorithms.cpp [hunk 3]
```diff
@@ -24,6 +24,98 @@
...
+    while (start_offset == 0 && start_node->parent()) {
+        start_offset = start_node->index();
+        start_node = start_node->parent();
+    }
+
+    // 5. If some inclusive ancestor of end node is an li, set end offset to one plus the index of the last such li in
+    //    tree order, and set end node to that li's parent.
+    ancestor = end_node;
+    do {
+        if (is<DOM::Element>(*ancestor) && static_cast<DOM::Element&>(*ancestor).local_name() == HTML::TagNames::li) {
```
Comment (tcl3) [inline Libraries/LibWeb/Editing/Internal/Algorithms.cpp:77]: You should be able to do `is<HTMLLIElement>()` here.

File: Libraries/LibWeb/Editing/Internal/Algorithms.cpp [hunk 4]
```diff
@@ -1566,4 +1692,194 @@ void split_the_parent_of_nodes(Vector<GC::Ref<DOM::Node>> const& nodes)
...
+        if (is_visible_node(node) || is<HTML::HTMLBRElement>(*node)) {
+            any_node_visible_or_br = true;
+            break;
+        }
+    }
+    if (!any_node_visible_or_br)
+        return {};
+
+    // 2. If node list's first member's parent is null, return null and abort these steps.
+    if (!node_list.first()->parent())
```
Comment (tcl3) [inline Libraries/LibWeb/Editing/Internal/Algorithms.cpp:1719]: Is `node_list` guaranteed to not be empty at this point?

File: Libraries/LibWeb/Editing/Commands.cpp [hunk 5]
```diff
@@ -384,9 +368,347 @@ bool command_delete_action(DOM::Document& document, String const&)
...
+            MUST(container->append_child(MUST(DOM::create_element(document, HTML::TagNames::br, Namespace::HTML))));
+
+        // 3. If container is a dd or dt, and it is not an allowed child of any of its ancestors in the same editing
+        //    host, set the tag name of container to the default single-line container name and let container be the
+        //    result.
+        if (static_cast<DOM::Element&>(*container).local_name().is_one_of(HTML::TagNames::dd, HTML::TagNames::dt)) {
+            bool allowed_child_of_any_ancestor = false;
+            GC::Ptr<DOM::Node> ancestor = container->parent();
+            do {
+                if (is_allowed_child_of_node(GC::Ref { *container }, GC::Ref { *ancestor })
```
[the remark this answers is not in the document]
Comment (gmta) [reply Libraries/LibWeb/Editing/Commands.cpp]: I've had some examples where I needed to be more defensive than the spec required, so I've changed all similar loops in the Editing API. Thanks!

File: Libraries/LibWeb/Editing/Internal/Algorithms.cpp (hunk 3 above)
Comment (gmta) [reply Libraries/LibWeb/Editing/Internal/Algorithms.cpp]: Good find! I've found a couple of other places too.

File: Libraries/LibWeb/Editing/Internal/Algorithms.cpp (hunk 4 above)
Comment (gmta) [reply Libraries/LibWeb/Editing/Internal/Algorithms.cpp:1723]: It is by virtue of all call sites always sending over at least one node, but that's a weak guarantee. I've added a `VERIFY`.

File: Libraries/LibWeb/Editing/Commands.cpp (hunk 1 above)
Comment (gmta) [reply Libraries/LibWeb/Editing/Commands.cpp:651]: Well, no? That's why there's a null check in the `while`?

File: Libraries/LibWeb/Editing/Commands.cpp [hunk 6]
```diff
@@ -384,9 +368,347 @@ bool command_delete_action(DOM::Document& document, String const&)
...
+    }
+
+    // 14. Let new line range be a new range whose start is the same as the active range's, and whose end is (container,
+    //     length of container).
+    auto new_line_range = DOM::Range::create(active_range.start_container(), active_range.start_offset(), *container, container->length());
+
+    // 15. While new line range's start offset is zero and its start node is not a prohibited paragraph child, set its
+    //     start to (parent of start node, index of start node).
+    while (new_line_range->start_offset() == 0 && !is_prohibited_paragraph_child(new_line_range->start_container()))
+        MUST(new_line_range->set_start(*new_line_range->start_container()->parent(), new_line_range->start_container()->index()));
```
[the remark this answers is not in the document]
Comment (gmta) [reply Libraries/LibWeb/Editing/Commands.cpp]: We can change the while condition to be a bit more defensive, for sure!

**Pull request merged**